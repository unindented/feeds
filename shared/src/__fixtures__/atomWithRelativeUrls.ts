/**
 * Copyright 2025 Daniel Perez Alvarez
 *
 * This file is part of Static Feeds.
 *
 * Static Feeds is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Static Feeds is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Static Feeds. If not, see <https://www.gnu.org/licenses/>.
 */

export const atomWithRelativeUrls = `<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
  <title></title>
  <subtitle>Recent content from Josh Ghent&#39;s blog</subtitle>
  <link href="https://joshghent.com/rss.xml" rel="self"/>
  <link href=""/>
  <updated>2025-01-09T00:00:00Z</updated>
  <id></id>
  <author>
    <name></name>
    <email></email>
  </author>
  
  <entry>
    <title>Event Driven Architecture Rules of Thumb</title>
    <link href="/eda-rules-of-thumb/"/>
    <updated>2025-01-09T00:00:00Z</updated>
    <id>/eda-rules-of-thumb/</id>
    <content type="html">&lt;p&gt;Event driven architectures are a fantastic mechanism for powering decoupled services. But they depend on the contract - the actual data points within each event.
As with these sorts of things, there is always an &amp;quot;it depends&amp;quot; of what data should go within events. Therefore, this guide is not prescriptive and should be followed as a &amp;quot;rules of thumb&amp;quot; - guiding principles, not rules.&lt;/p&gt;</content>
  </entry>
  
  <entry>
    <title>ClickOps and IaC - which are Pets and Cattle?</title>
    <link href="/pets-vs-cattle/"/>
    <updated>2025-01-08T00:00:00Z</updated>
    <id>/pets-vs-cattle/</id>
    <content type="html">&lt;p&gt;Ancient history tells us that a peoples known as the &amp;quot;sysadmins&amp;quot; or &amp;quot;web masters&amp;quot; used to manually configure servers via the archaic command line. These people literally SSH&#39;ed into machines and ran commands on them. Utter animals.&lt;/p&gt;</content>
  </entry>
</feed>
`;
