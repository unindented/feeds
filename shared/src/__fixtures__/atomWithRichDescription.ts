/**
 * Copyright 2025 Daniel Perez Alvarez
 *
 * This file is part of Static Feeds.
 *
 * Static Feeds is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Static Feeds is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Static Feeds. If not, see <https://www.gnu.org/licenses/>.
 */

export const atomWithRichDescription = `<?xml version="1.0" encoding="utf-8"?><feed xmlns="http://www.w3.org/2005/Atom">
  <title>fasterthanli.me</title>
  <subtitle>amos likes to tinker</subtitle>
  <link href="https://fasterthanli.me"/>
  <link rel="self" href="https://fasterthanli.me/index.xml"/>
  <id>https://fasterthanli.me/</id>
  <updated>2025-02-07T18:53:01Z</updated>
  <author>
    <name>Amos Wenger</name>
    <uri>https://fasterthanli.me</uri>
  </author>
  <category term="programming"/>
  <category term="computers"/>
  <category term="rust"/>
  <generator>told</generator>
  <icon>https://cdn.fasterthanli.me/content/img/logo-square-2~de9d25724fcb4af8.avif</icon>

  <entry>
    <title></title>
    <link rel="alternate" href="https://fasterthanli.me/articles/the-case-for-sans-io"/>
    <id>https://fasterthanli.me/articles/the-case-for-sans-io</id>
    <updated>2025-02-07T18:53:01Z</updated>
    <summary type="html">&lt;p data-bo=&quot;208&quot;&gt;The most popular option to decompress ZIP files from the Rust programming
language is a crate simply named &lt;a href=&quot;https:&#x2f;&#x2f;lib.rs&#x2f;crates&#x2f;zip&quot; id=&quot;zip&quot;&gt;zip&lt;&#x2f;a&gt; — At the time of this writing, it has 48
million downloads. It’s fully-featured, supporting various compression methods,
encryption, and even supports writing zip files.&lt;&#x2f;p&gt;</summary>
    <category term="zip"/>
    <category term="rust"/>
    <category term="async"/>
  </entry>

  <entry>
    <title>Catching up with async Rust</title>
    <link rel="alternate" href="https://fasterthanli.me/articles/catching-up-with-async-rust"/>
    <id>https://fasterthanli.me/articles/catching-up-with-async-rust</id>
    <updated>2024-12-25T07:30:00Z</updated>
    <summary type="html">&lt;p data-bo=&quot;211&quot;&gt;In December 2023, a minor miracle happened: &lt;a href=&quot;https:&#x2f;&#x2f;blog.rust-lang.org&#x2f;2023&#x2f;12&#x2f;21&#x2f;async-fn-rpit-in-traits.html&quot;&gt;async fn in traits&lt;&#x2f;a&gt; shipped.&lt;&#x2f;p&gt;

&lt;p data-bo=&quot;354&quot;&gt;As of Rust 1.39, we already had free-standing async functions:&lt;&#x2f;p&gt;

&lt;code class=&quot;code-block&quot; translate=&quot;no&quot; data-lang=&quot;rust&quot; data-bo=&quot;418&quot;&gt;&lt;pre class=&quot;scroll-wrapper&quot;&gt;&lt;i class=hh4&gt;pub&lt;&#x2f;i&gt; &lt;i class=hh4&gt;async&lt;&#x2f;i&gt; &lt;i class=hh4&gt;fn&lt;&#x2f;i&gt; &lt;i class=hh3&gt;read_hosts&lt;&#x2f;i&gt;&lt;i class=hh8&gt;(&lt;&#x2f;i&gt;&lt;i class=hh8&gt;)&lt;&#x2f;i&gt; -&amp;gt; eyre&lt;i class=hh9&gt;::&lt;&#x2f;i&gt;&lt;i class=hh13&gt;Result&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;lt;&lt;&#x2f;i&gt;&lt;i class=hh13&gt;Vec&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;lt;&lt;&#x2f;i&gt;&lt;i class=hh14&gt;u8&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;gt;&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;gt;&lt;&#x2f;i&gt; &lt;i class=hh8&gt;{&lt;&#x2f;i&gt;
  &lt;i class=hh18&gt;&#x2f;&#x2f; etc.&lt;&#x2f;i&gt;
&lt;i class=hh8&gt;}&lt;&#x2f;i&gt;
&lt;&#x2f;pre&gt;&lt;&#x2f;code&gt;&lt;p data-bo=&quot;498&quot;&gt;…and async functions in impl blocks:&lt;&#x2f;p&gt;

&lt;code class=&quot;code-block&quot; translate=&quot;no&quot; data-lang=&quot;rust&quot; data-bo=&quot;538&quot;&gt;&lt;pre class=&quot;scroll-wrapper&quot;&gt;&lt;i class=hh4&gt;impl&lt;&#x2f;i&gt; &lt;i class=hh13&gt;HostReader&lt;&#x2f;i&gt; &lt;i class=hh8&gt;{&lt;&#x2f;i&gt;
  &lt;i class=hh4&gt;pub&lt;&#x2f;i&gt; &lt;i class=hh4&gt;async&lt;&#x2f;i&gt; &lt;i class=hh4&gt;fn&lt;&#x2f;i&gt; &lt;i class=hh3&gt;read_hosts&lt;&#x2f;i&gt;&lt;i class=hh8&gt;(&lt;&#x2f;i&gt;&lt;i class=hh5&gt;&amp;amp;&lt;&#x2f;i&gt;&lt;i class=hh16&gt;self&lt;&#x2f;i&gt;&lt;i class=hh8&gt;)&lt;&#x2f;i&gt; -&amp;gt; eyre&lt;i class=hh9&gt;::&lt;&#x2f;i&gt;&lt;i class=hh13&gt;Result&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;lt;&lt;&#x2f;i&gt;&lt;i class=hh13&gt;Vec&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;lt;&lt;&#x2f;i&gt;&lt;i class=hh14&gt;u8&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;gt;&lt;&#x2f;i&gt;&lt;i class=hh8&gt;&amp;gt;&lt;&#x2f;i&gt; &lt;i class=hh8&gt;{&lt;&#x2f;i&gt;
    &lt;i class=hh18&gt;&#x2f;&#x2f; etc.&lt;&#x2f;i&gt;
  &lt;i class=hh8&gt;}&lt;&#x2f;i&gt;
&lt;i class=hh8&gt;}&lt;&#x2f;i&gt;
&lt;&#x2f;pre&gt;&lt;&#x2f;code&gt;&lt;p data-bo=&quot;655&quot;&gt;But we did not have async functions in traits:&lt;&#x2f;p&gt;
</summary>
    <category term="rust"/>
    <category term="async"/>
    <category term="traits"/>
  </entry>
</feed>`;
