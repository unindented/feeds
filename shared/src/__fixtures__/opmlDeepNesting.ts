/**
 * Copyright 2025 Daniel Perez Alvarez
 *
 * This file is part of Static Feeds.
 *
 * Static Feeds is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Static Feeds is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Static Feeds. If not, see <https://www.gnu.org/licenses/>.
 */

export const opmlDeepNesting = `<?xml version="1.0" encoding="UTF-8"?>
<opml version="1.1">
  <head>
    <title>My feeds</title>
  </head>
  <body>
    <outline text="Technology">
      <outline text="Web Development">
        <outline text="Lea Verou" type="rss" htmlUrl="https://lea.verou.me/" xmlUrl="https://lea.verou.me/feed.xml"/>
        <outline text="Accessibility">
          <outline text="LÃ©onie Watson" type="rss" htmlUrl="https://tink.uk/" xmlUrl="https://tink.uk/feed.xml"/>
        </outline>
      </outline>
    </outline>
  </body>
</opml>
`;
