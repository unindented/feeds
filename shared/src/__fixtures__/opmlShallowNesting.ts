/**
 * Copyright 2025 Daniel Perez Alvarez
 *
 * This file is part of Static Feeds.
 *
 * Static Feeds is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Static Feeds is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Static Feeds. If not, see <https://www.gnu.org/licenses/>.
 */

export const opmlShallowNesting = `<?xml version="1.0" encoding="UTF-8"?>
<opml version="1.1">
  <head>
    <title>My feeds</title>
  </head>
  <body>
    <outline text="Rust">
      <outline text="Amos Wenger" type="rss" htmlUrl="https://fasterthanli.me/" xmlUrl="https://fasterthanli.me/index.xml"/>
      <outline text="Rain" type="rss" htmlUrl="https://sunshowers.io/" xmlUrl="https://sunshowers.io/index.xml"/>
    </outline>
  </body>
</opml>
`;
