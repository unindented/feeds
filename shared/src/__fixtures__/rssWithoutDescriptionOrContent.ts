/**
 * Copyright 2025 Daniel Perez Alvarez
 *
 * This file is part of Static Feeds.
 *
 * Static Feeds is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Static Feeds is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Static Feeds. If not, see <https://www.gnu.org/licenses/>.
 */

export const rssWithoutDescriptionOrContent = `<?xml version="1.0"?>
<rss version="2.0" xmlns:atom="http://www.w3.org/2005/Atom">
  <channel> 
    <title>Fabien Sanglard</title>
    <atom:link href="https://fabiensanglard.net/rss.xml" rel="self" type="application/rss+xml" />
    <link>https://fabiensanglard.net</link>
    <description>Chronicles of software wizardry</description>
    <language>en-us</language>

    <item>
      <title>Watching sunsets</title>
      <guid>https://fabiensanglard.net/sunset/index.html</guid>
      <link>https://fabiensanglard.net/sunset/index.html</link>
      <pubDate>18 Aug 2024 00:00:00 +0000</pubDate>
    </item>

    <item>
      <title>SNES: Sprites and backgrounds rendering</title>
      <guid>https://fabiensanglard.net/snes_ppus_why/index.html</guid>
      <link>https://fabiensanglard.net/snes_ppus_why/index.html</link>
      <pubDate>09 Aug 2024 00:00:00 +0000</pubDate>
    </item>
 </channel>
</rss>`;
