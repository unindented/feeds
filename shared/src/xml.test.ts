/**
 * Copyright 2025 Daniel Perez Alvarez
 *
 * This file is part of Static Feeds.
 *
 * Static Feeds is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, either version 3 of the License, or (at your option) any
 * later version.
 *
 * Static Feeds is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Static Feeds. If not, see <https://www.gnu.org/licenses/>.
 */

import { fs, vol } from "memfs";
import type { MockInstance } from "vitest";
import { beforeEach, describe, expect, it, vi } from "vitest";

import { atomWithRelativeUrls } from "./__fixtures__/atomWithRelativeUrls.ts";
import { atomWithRichDescription } from "./__fixtures__/atomWithRichDescription.ts";
import { opmlDeepNesting } from "./__fixtures__/opmlDeepNesting.ts";
import { opmlNoNesting } from "./__fixtures__/opmlNoNesting.ts";
import { opmlShallowNesting } from "./__fixtures__/opmlShallowNesting.ts";
import { opmlUntitledOutlines } from "./__fixtures__/opmlUntitledOutlines.ts";
import { rssWithDescriptionAndContent } from "./__fixtures__/rssWithDescriptionAndContent.ts";
import { rssWithoutDescriptionOrContent } from "./__fixtures__/rssWithoutDescriptionOrContent.ts";
import { rssWithPlainDescription } from "./__fixtures__/rssWithPlainDescription.ts";
import { rssWithRichDescription } from "./__fixtures__/rssWithRichDescription.ts";
import type { SimpleFeed } from "./types.ts";
import { flattenFeeds, readFeed, readOpmlFile, unflattenFeeds } from "./xml.ts";

vi.mock("node:fs");
vi.mock("node:fs/promises");

interface XmlTestEachColumns {
  name: string;
  sourceContent: string;
}

describe("readOpmlFile", () => {
  beforeEach(() => {
    vol.reset();
  });

  it.each`
    name                                                  | sourceContent
    ${"parses an OPML file without nested outlines"}      | ${opmlNoNesting}
    ${"parses an OPML file with shallow nested outlines"} | ${opmlShallowNesting}
    ${"parses an OPML file with deep nested outlines"}    | ${opmlDeepNesting}
    ${"parses an OPML file with untitled outlines"}       | ${opmlUntitledOutlines}
  `("$name", async ({ sourceContent }: XmlTestEachColumns) => {
    const path = "/foobar.opml";
    await fs.promises.writeFile(path, sourceContent);

    await expect(readOpmlFile(path)).resolves.toMatchSnapshot();
  });
});

describe("flattenFeeds", () => {
  beforeEach(() => {
    vol.reset();
  });

  it.each`
    name                                                                  | sourceContent
    ${"gathers feeds from an OPML document without nested outlines"}      | ${opmlNoNesting}
    ${"gathers feeds from an OPML document with shallow nested outlines"} | ${opmlShallowNesting}
    ${"gathers feeds from an OPML document with deep nested outlines"}    | ${opmlDeepNesting}
    ${"gathers feeds from an OPML document with untitled outlines"}       | ${opmlUntitledOutlines}
  `("$name", async ({ sourceContent }: XmlTestEachColumns) => {
    const path = "/foobar.opml";
    await fs.promises.writeFile(path, sourceContent);

    const opmlDocument = await readOpmlFile(path);

    expect(flattenFeeds(opmlDocument)).toMatchSnapshot();
  });
});

describe("unflattenFeeds", () => {
  beforeEach(() => {
    vol.reset();
  });

  it.each`
    name                                                        | sourceContent
    ${"rebuilds an OPML document without nested outlines"}      | ${opmlNoNesting}
    ${"rebuilds an OPML document with shallow nested outlines"} | ${opmlShallowNesting}
    ${"rebuilds an OPML document with deep nested outlines"}    | ${opmlDeepNesting}
  `("$name", async ({ sourceContent }: XmlTestEachColumns) => {
    const path = "/foobar.opml";
    await fs.promises.writeFile(path, sourceContent);

    const opmlDocument = await readOpmlFile(path);

    expect(unflattenFeeds(flattenFeeds(opmlDocument)).opml.body).toStrictEqual(
      opmlDocument.opml.body,
    );
  });
});

describe("readFeed", () => {
  let fetchSpy: MockInstance<typeof global.fetch>;

  beforeEach(() => {
    fetchSpy = vi.spyOn(global, "fetch");
  });

  it.each`
    name                                                             | sourceContent
    ${"reads an RSS feed with items without description or content"} | ${rssWithoutDescriptionOrContent}
    ${"reads an RSS feed with items with plain description"}         | ${rssWithPlainDescription}
    ${"reads an RSS feed with items with rich description"}          | ${rssWithRichDescription}
    ${"reads an RSS feed with items with description and content"}   | ${rssWithDescriptionAndContent}
    ${"reads an Atom feed with items with rich description"}         | ${atomWithRichDescription}
    ${"reads an Atom feed with relative URLs"}                       | ${atomWithRelativeUrls}
  `("$name", async ({ sourceContent }: XmlTestEachColumns) => {
    fetchSpy.mockResolvedValue(new Response(sourceContent));

    const feed: SimpleFeed = {
      id: "example-com",
      htmlUrl: "https://example.com",
      xmlUrl: "https://example.com/index.xml",
      title: "Example",
      path: null,
    };

    await expect(readFeed(feed)).resolves.toMatchSnapshot();
  });
});
