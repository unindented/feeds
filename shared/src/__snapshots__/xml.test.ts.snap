// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`flattenFeeds > 'gathers feeds from an OPML document w…' 1`] = `
[
  {
    "htmlUrl": "https://nicholas.carlini.com/",
    "id": "nicholas-carlini-com",
    "path": null,
    "title": "Nicholas Carlini",
    "xmlUrl": "https://nicholas.carlini.com/writing/feed.xml",
  },
  {
    "htmlUrl": "https://simonwillison.net/",
    "id": "simonwillison-net",
    "path": null,
    "title": "Simon Willison",
    "xmlUrl": "https://simonwillison.net/atom/everything/",
  },
]
`;

exports[`flattenFeeds > 'gathers feeds from an OPML document w…' 2`] = `
[
  {
    "htmlUrl": "https://fasterthanli.me/",
    "id": "fasterthanli-me",
    "path": "Rust",
    "title": "Amos Wenger",
    "xmlUrl": "https://fasterthanli.me/index.xml",
  },
  {
    "htmlUrl": "https://sunshowers.io/",
    "id": "sunshowers-io",
    "path": "Rust",
    "title": "Rain",
    "xmlUrl": "https://sunshowers.io/index.xml",
  },
]
`;

exports[`flattenFeeds > 'gathers feeds from an OPML document w…' 3`] = `
[
  {
    "htmlUrl": "https://lea.verou.me/",
    "id": "lea-verou-me",
    "path": "Technology/Web Development",
    "title": "Lea Verou",
    "xmlUrl": "https://lea.verou.me/feed.xml",
  },
  {
    "htmlUrl": "https://tink.uk/",
    "id": "tink-uk",
    "path": "Technology/Web Development/Accessibility",
    "title": "Léonie Watson",
    "xmlUrl": "https://tink.uk/feed.xml",
  },
]
`;

exports[`flattenFeeds > 'gathers feeds from an OPML document w…' 4`] = `
[
  {
    "htmlUrl": "https://arstechnica.com/",
    "id": "arstechnica-com",
    "path": "News/Technology",
    "title": "Ars Technica",
    "xmlUrl": "https://arstechnica.com/feed/",
  },
  {
    "htmlUrl": "https://www.theverge.com/",
    "id": "theverge-com",
    "path": "News/Untitled",
    "title": "Untitled",
    "xmlUrl": "https://www.theverge.com/rss/index.xml",
  },
]
`;

exports[`readFeed > 'reads an Atom feed with items with ri…' 1`] = `
[
  {
    "feedId": "example-com",
    "htmlUrl": "https://fasterthanli.me/articles/the-case-for-sans-io",
    "id": "https://fasterthanli.me/articles/the-case-for-sans-io",
    "published": 2025-02-07T18:53:01.000Z,
    "summary": "The most popular option to decompress ZIP files from the Rust programming language is a crate simply named zip — At the time of this writing, it has 48 million downloads. It’s fully-featured, supporting various compression methods, encryption, and even supports writing zip files.",
    "title": "Untitled",
    "updated": 2025-02-07T18:53:01.000Z,
  },
  {
    "feedId": "example-com",
    "htmlUrl": "https://fasterthanli.me/articles/catching-up-with-async-rust",
    "id": "https://fasterthanli.me/articles/catching-up-with-async-rust",
    "published": 2024-12-25T07:30:00.000Z,
    "summary": "In December 2023, a minor miracle happened: async fn in traits shipped. As of Rust 1.39, we already had free-standing async functions: pub async fn read_hosts() -> eyre::ResultVecu8>> { // etc. } …and async functions in impl blocks: impl HostReader { pub async fn read_hosts(&self) -> eyre::ResultVecu8…",
    "title": "Catching up with async Rust",
    "updated": 2024-12-25T07:30:00.000Z,
  },
]
`;

exports[`readFeed > 'reads an Atom feed with relative URLs' 1`] = `
[
  {
    "feedId": "example-com",
    "htmlUrl": "https://joshghent.com/eda-rules-of-thumb/",
    "id": "https://joshghent.com/eda-rules-of-thumb/",
    "published": 2025-01-09T00:00:00.000Z,
    "summary": "Event driven architectures are a fantastic mechanism for powering decoupled services. But they depend on the contract - the actual data points within each event. As with these sorts of things, there is always an "it depends" of what data should go within events. Therefore, this guide is not prescriptive…",
    "title": "Event Driven Architecture Rules of Thumb",
    "updated": 2025-01-09T00:00:00.000Z,
  },
  {
    "feedId": "example-com",
    "htmlUrl": "https://joshghent.com/pets-vs-cattle/",
    "id": "https://joshghent.com/pets-vs-cattle/",
    "published": 2025-01-08T00:00:00.000Z,
    "summary": "Ancient history tells us that a peoples known as the "sysadmins" or "web masters" used to manually configure servers via the archaic command line. These people literally SSH'ed into machines and ran commands on them. Utter animals.",
    "title": "ClickOps and IaC - which are Pets and Cattle?",
    "updated": 2025-01-08T00:00:00.000Z,
  },
]
`;

exports[`readFeed > 'reads an RSS feed with items with des…' 1`] = `
[
  {
    "feedId": "example-com",
    "htmlUrl": "https://krebsonsecurity.com/2025/02/how-phished-data-turns-into-apple-google-wallets/",
    "id": "https://krebsonsecurity.com/2025/02/how-phished-data-turns-into-apple-google-wallets/",
    "published": 2025-02-18T18:37:26.000Z,
    "summary": "Carding -- the underground business of stealing, selling and swiping stolen payment card data -- has long been the dominion of Russia-based hackers. Happily, the broad deployment of more secure chip-based payment cards in the United States has weakened the carding market. But a flurry of innovation from…",
    "title": "How Phished Data Turns into Apple & Google Wallets",
    "updated": 2025-02-18T18:37:26.000Z,
  },
  {
    "feedId": "example-com",
    "htmlUrl": "https://krebsonsecurity.com/2025/02/nearly-a-year-later-mozilla-is-still-promoting-onerep/",
    "id": "https://krebsonsecurity.com/2025/02/nearly-a-year-later-mozilla-is-still-promoting-onerep/",
    "published": 2025-02-13T20:14:47.000Z,
    "summary": "In mid-March 2024, KrebsOnSecurity revealed that the founder of the personal data removal service Onerep also founded dozens of people-search companies. Shortly after that investigation was published, Mozilla said it would stop bundling Onerep with the Firefox browser and wind down its partnership. But…",
    "title": "Nearly a Year Later, Mozilla is Still Promoting OneRep",
    "updated": 2025-02-13T20:14:47.000Z,
  },
]
`;

exports[`readFeed > 'reads an RSS feed with items with pla…' 1`] = `
[
  {
    "feedId": "example-com",
    "htmlUrl": "https://nicholas.carlini.com/writing/2025/forecasting-ai-2025-update.html",
    "id": "https://nicholas.carlini.com/writing/2025/forecasting-ai-2025-update.html",
    "published": 2025-02-10T00:00:00.000Z,
    "summary": "Late last year, I published an article asking readers to make 30 forecasts about the future of AI in 2027 and 2030—from whether or not you could buy a robot to do your laundry, to predicting the valuation of leading AI labs, to estimating the likelihood of an AI-caused catastrophe.",
    "title": "AI forecasting retrospective: you're (probably) over-confident",
    "updated": 2025-02-10T00:00:00.000Z,
  },
  {
    "feedId": "example-com",
    "htmlUrl": "https://nicholas.carlini.com/writing/2025/regex-chess.html",
    "id": "https://nicholas.carlini.com/writing/2025/regex-chess.html",
    "published": 2025-01-06T00:00:00.000Z,
    "summary": "Over the holidays I decided it's been too long since I did something with entirely no purpose. So without further ado, I present to you ... Regex Chess: sequence of 84,688 regular expressions that, when executed in order, will play a (valid; not entirely terrible) move given a chess board as input.",
    "title": "Regex Chess: A 2-ply minimax chess engine in 84,688 regular expressions",
    "updated": 2025-01-06T00:00:00.000Z,
  },
]
`;

exports[`readFeed > 'reads an RSS feed with items with ric…' 1`] = `
[
  {
    "feedId": "example-com",
    "htmlUrl": "https://sunshowers.io/posts/monads-through-pbt/",
    "id": "https://sunshowers.io/posts/monads-through-pbt/",
    "published": 2025-02-19T00:00:00.000Z,
    "summary": "In programming pedagogy, monads have a place as a mystical object from the functional programming world that’s hard to understand and even harder to explain. The stereotype about monad explanations is that they fall into two buckets: either comparisons to some kind of food item, or throwing complex mathematical…",
    "title": "Demystifying monads in Rust through property-based testing",
    "updated": 2025-02-19T00:00:00.000Z,
  },
  {
    "feedId": "example-com",
    "htmlUrl": "https://sunshowers.io/posts/there-is-no-free-will/",
    "id": "https://sunshowers.io/posts/there-is-no-free-will/",
    "published": 2025-01-29T00:00:00.000Z,
    "summary": "There’s absolutely no brand-new insight in this post. A huge amount of credit goes to various philosophers and thinkers, especially folks like Aaron Rabinowitz, for shaping my views. Any errors in it are my own. In the natural world, everything that occurs is the result of a chain of causation all the…",
    "title": "Free will quite clearly doesn't exist",
    "updated": 2025-01-29T00:00:00.000Z,
  },
]
`;

exports[`readFeed > 'reads an RSS feed with items without …' 1`] = `
[
  {
    "feedId": "example-com",
    "htmlUrl": "https://fabiensanglard.net/sunset/index.html",
    "id": "https://fabiensanglard.net/sunset/index.html",
    "published": 2024-08-18T00:00:00.000Z,
    "summary": null,
    "title": "Watching sunsets",
    "updated": 2024-08-18T00:00:00.000Z,
  },
  {
    "feedId": "example-com",
    "htmlUrl": "https://fabiensanglard.net/snes_ppus_why/index.html",
    "id": "https://fabiensanglard.net/snes_ppus_why/index.html",
    "published": 2024-08-09T00:00:00.000Z,
    "summary": null,
    "title": "SNES: Sprites and backgrounds rendering",
    "updated": 2024-08-09T00:00:00.000Z,
  },
]
`;

exports[`readOpmlFile > 'parses an OPML file with deep nested …' 1`] = `
{
  "?xml": {
    "encoding": "UTF-8",
    "version": "1.0",
  },
  "opml": {
    "body": {
      "outline": [
        {
          "outline": [
            {
              "outline": [
                {
                  "htmlUrl": "https://lea.verou.me/",
                  "text": "Lea Verou",
                  "type": "rss",
                  "xmlUrl": "https://lea.verou.me/feed.xml",
                },
                {
                  "outline": [
                    {
                      "htmlUrl": "https://tink.uk/",
                      "text": "Léonie Watson",
                      "type": "rss",
                      "xmlUrl": "https://tink.uk/feed.xml",
                    },
                  ],
                  "text": "Accessibility",
                },
              ],
              "text": "Web Development",
            },
          ],
          "text": "Technology",
        },
      ],
    },
    "head": {
      "title": "My feeds",
    },
    "version": "1.1",
  },
}
`;

exports[`readOpmlFile > 'parses an OPML file with shallow nest…' 1`] = `
{
  "?xml": {
    "encoding": "UTF-8",
    "version": "1.0",
  },
  "opml": {
    "body": {
      "outline": [
        {
          "outline": [
            {
              "htmlUrl": "https://fasterthanli.me/",
              "text": "Amos Wenger",
              "type": "rss",
              "xmlUrl": "https://fasterthanli.me/index.xml",
            },
            {
              "htmlUrl": "https://sunshowers.io/",
              "text": "Rain",
              "type": "rss",
              "xmlUrl": "https://sunshowers.io/index.xml",
            },
          ],
          "text": "Rust",
        },
      ],
    },
    "head": {
      "title": "My feeds",
    },
    "version": "1.1",
  },
}
`;

exports[`readOpmlFile > 'parses an OPML file with untitled out…' 1`] = `
{
  "?xml": {
    "encoding": "UTF-8",
    "version": "1.0",
  },
  "opml": {
    "body": {
      "outline": [
        {
          "outline": [
            {
              "outline": [
                {
                  "htmlUrl": "https://arstechnica.com/",
                  "text": "Ars Technica",
                  "type": "rss",
                  "xmlUrl": "https://arstechnica.com/feed/",
                },
              ],
              "text": "Technology",
            },
            {
              "outline": [
                {
                  "htmlUrl": "https://www.theverge.com/",
                  "type": "rss",
                  "xmlUrl": "https://www.theverge.com/rss/index.xml",
                },
              ],
            },
          ],
          "text": "News",
        },
      ],
    },
    "head": {
      "title": "My feeds",
    },
    "version": "1.1",
  },
}
`;

exports[`readOpmlFile > 'parses an OPML file without nested ou…' 1`] = `
{
  "?xml": {
    "encoding": "UTF-8",
    "version": "1.0",
  },
  "opml": {
    "body": {
      "outline": [
        {
          "htmlUrl": "https://nicholas.carlini.com/",
          "text": "Nicholas Carlini",
          "type": "rss",
          "xmlUrl": "https://nicholas.carlini.com/writing/feed.xml",
        },
        {
          "htmlUrl": "https://simonwillison.net/",
          "text": "Simon Willison",
          "type": "rss",
          "xmlUrl": "https://simonwillison.net/atom/everything/",
        },
      ],
    },
    "head": {
      "title": "My feeds",
    },
    "version": "1.1",
  },
}
`;
